### ISSUE TITLE
Unexpected error message when using eval

### ISSUE THREAD

**Issue by bpinto** [fish-shell/fish-shell#2648]
```
### ~/.config/fish/functions/eval_bug.fish

function inner_function
  echo 'abc'
end

function eval_bug
  eval "inner_function >/dev/null"
end
```

```
$ fish -c "eval_bug"
An error occurred while redirecting file descriptor 1
dup2: Bad file descriptor
```

I have read the [eval](https://github.com/fish-shell/fish-shell/blob/master/share/functions/eval.fish) file comments a couple times but I can't say I understood it. :P

Is this a bug? Or is it expected?

---

Debugging information:
fish, version 2.2.0-436-gef67fc7
OSX El Capitan

This has been reproduced on Ubuntu machines however.

**Comment by krader1961** [comment]
It looks like a bug to me. You can workaround the problem by moving the redirection out of the eval'd string. It's also worth noting that if you define the inner function in an interactive shell then run the eval with the redirection inside the eval'd string it works as expected.

**Comment by bpinto** [comment]
Thanks @krader1961. I have found this while doing tests with fish-spec and the workaround I have used is to assign the value to a variable instead.

I dont understand what happens there to be able to work on a PR. 😕

**Comment by krader1961** [comment]
This seems like a relatively obscure bug that will affect very few people and doesn't need to hold up the 2.3.0 release.

**Comment by drewcassidy** [comment]
I'm getting this whenever fish starts after I added a function to alias the github `hub` tool

**Comment by ridiculousfish** [comment]
This no longer seems to reproduce for me on Ubuntu or Mac, testing on 9652b3e11b39d8fd56f565d39059fb419fd95fc4. Closing, please comment if you still see this.

**Comment by zanchey** [comment]
Fixed as far back as 2.7 (don't have any other versions handy), so I will set it to that milestone.
